File: OpenCodeLauncher/Services/OpenCodeUpdateService.cs
using System.Diagnostics;
using System.IO;
using System.Text.Json;

namespace OpenCodeLauncher.Services;

public sealed record OpenCodeUpdateResult(string Status, string Message, int ExitCode);

public sealed class OpenCodeUpdateService
{
    public async Task<OpenCodeUpdateResult> CheckAsync()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var script = Path.Combine(home, "proggs", "opencode-setup", "build-install-windows-mousefix.ps1");
        if (!File.Exists(script)) return new("unavailable", "Update-Skript nicht gefunden.", 0);

        var psi = new ProcessStartInfo
        {
            FileName = "pwsh.exe",
            UseShellExecute = false,
            CreateNoWindow = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            WorkingDirectory = Path.GetDirectoryName(script)!
        };
        psi.ArgumentList.Add("-NoLogo");
        psi.ArgumentList.Add("-NoProfile");
        psi.ArgumentList.Add("-ExecutionPolicy");
        psi.ArgumentList.Add("Bypass");
        psi.ArgumentList.Add("-File");
        psi.ArgumentList.Add(script);

        try
        {
            using var process = Process.Start(psi) ?? throw new InvalidOperationException("Update-Prozess konnte nicht gestartet werden.");
            var stdoutTask = process.StandardOutput.ReadToEndAsync();
            var stderrTask = process.StandardError.ReadToEndAsync();
            await process.WaitForExitAsync();
            var stdout = (await stdoutTask).Trim();
            var stderr = (await stderrTask).Trim();
            var state = ReadState(home);
            var status = stdout.StartsWith("OPENCODE_UPDATE_STATUS=deferred", StringComparison.Ordinal)
                ? "deferred"
                : stdout.StartsWith("OPENCODE_UPDATE_STATUS=busy", StringComparison.Ordinal)
                    ? "busy"
                    : state.Status ?? (process.ExitCode == 0 ? "completed" : "failed");
            var message = status == "deferred"
                ? "Die tägliche OpenCode-Updateprüfung ist noch nicht fällig."
                : status == "busy"
                    ? "Eine OpenCode-Updateprüfung läuft bereits."
                : state.Message ?? (process.ExitCode == 0 ? stdout : stderr);
            Logger.Instance.Info("OpenCodeUpdateService", "CheckAsync", "Windows-Fix-Update abgeschlossen",
                new { status, message, process.ExitCode, stdout, stderr });
            return new(status, message, process.ExitCode);
        }
        catch (Exception ex)
        {
            Logger.Instance.Warn("OpenCodeUpdateService", "CheckAsync", $"Updateprüfung nicht verfügbar: {ex.Message}");
            return new("failed", ex.Message, -1);
        }
    }

    private static (string? Status, string? Message) ReadState(string home)
    {
        var path = Path.Combine(home, ".local", "share", "opencode-mousefix", "update-state.json");
        try
        {
            using var json = JsonDocument.Parse(File.ReadAllText(path));
            var root = json.RootElement;
            return (
                root.TryGetProperty("status", out var status) ? status.GetString() : null,
                root.TryGetProperty("message", out var message) ? message.GetString() : null
            );
        }
        catch (Exception ex)
        {
            Logger.Instance.Warn("OpenCodeUpdateService", "ReadState", $"Update-State konnte nicht gelesen werden: {ex.Message}");
            return (null, null);
        }
    }
}

File: OpenLauncherMac/OpenLauncher/Services/OpenCodeUpdateService.swift
import Foundation

struct OpenCodeUpdateResult {
    let status: String
    let message: String
    let exitCode: Int32
}

/// Taegliche OpenCode-Updatepruefung beim Start des Launchers.
/// 1:1-Port von Services/OpenCodeUpdateService.cs - mit einem bewussten Plattform-Unterschied:
///
/// Windows startet `opencode-setup/build-install-windows-mousefix.ps1`. Dieses Skript baut einen
/// gepatchten opencode-Build gegen einen **Windows-spezifischen Maus-Fehler**; auf macOS gibt es
/// weder den Fehler noch den Build. Der Mac sucht deshalb ein macOS-Update-Skript
/// (`opencode-setup/update-macos.sh`) und meldet, wenn es fehlt, denselben Zustand wie Windows bei
/// fehlendem Skript: "unavailable". Dieser Zustand veraendert die Statuszeile nicht - der Launcher
/// startet also unveraendert, statt eine Fehlermeldung zu zeigen.
///
/// Bewusst wird NICHT `opencode-setup/install.sh` aufgerufen: das ist ein vollstaendiger
/// Einrichtungslauf, der ~/.config/opencode ueberschreibt - kein Update-Check.
final class OpenCodeUpdateService {
    func check() async -> OpenCodeUpdateResult {
        let script = (Paths.home as NSString).appendingPathComponent("proggs/opencode-setup/update-macos.sh")
        guard Paths.fileExists(script) else {
            return OpenCodeUpdateResult(status: "unavailable", message: "Update-Skript nicht gefunden.", exitCode: 0)
        }

        let workingDirectory = (script as NSString).deletingLastPathComponent
        let result = await Task.detached {
            Shell.run("/bin/bash", [script], workingDirectory: workingDirectory, timeout: 120)
        }.value

        let stdout = result.stdout.trimmingCharacters(in: .whitespacesAndNewlines)
        let stderr = result.stderr.trimmingCharacters(in: .whitespacesAndNewlines)
        let state = readState()

        let status: String
        if stdout.hasPrefix("OPENCODE_UPDATE_STATUS=deferred") {
            status = "deferred"
        } else if stdout.hasPrefix("OPENCODE_UPDATE_STATUS=busy") {
            status = "busy"
        } else {
            status = state.status ?? (result.exitCode == 0 ? "completed" : "failed")
        }

        let message: String
        switch status {
        case "deferred": message = "Die tägliche OpenCode-Updateprüfung ist noch nicht fällig."
        case "busy": message = "Eine OpenCode-Updateprüfung läuft bereits."
        default: message = state.message ?? (result.exitCode == 0 ? stdout : stderr)
        }

        Logger.shared.info("OpenCodeUpdateService", "check", "OpenCode-Updatepruefung abgeschlossen",
                           ["status": status, "message": message, "exitCode": Int(result.exitCode)])
        return OpenCodeUpdateResult(status: status, message: message, exitCode: result.exitCode)
    }

    private func readState() -> (status: String?, message: String?) {
        let path = (Paths.home as NSString).appendingPathComponent(".local/share/opencode-mousefix/update-state.json")
        guard let data = FileManager.default.contents(atPath: path),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return (nil, nil)
        }
        return (root["status"] as? String, root["message"] as? String)
    }
}
